superset-frontend/src/dashboard/util/getFilterConfigsFromFormdata.js

```diff
@@ -20,89 +20,85 @@
 import { TIME_FILTER_MAP } from '../../visualizations/FilterBox/FilterBox';
 import { TIME_FILTER_LABELS } from '../../explore/constants';
 
-export default function getFilterConfigsFromFormdata(form_data = {}) {
+/**
+ * Parse filters for Table chart. All non-metric columns are considered
+ * filterable values.
+ */
+function getFilterConfigsFromTableChart(form_data = {}) {
+  const { groupby = [], all_columns = [] } = form_data;
+  const configs = { columns: {}, labels: {} };
+  // `groupby` is from GROUP BY mode (aggregations)
+  // `all_columns` is from NOT GROUP BY mode (raw records)
+  const columns = groupby.concat(all_columns);
+  columns.forEach(column => {
+    configs.columns[column] = undefined;
+    configs.labels[column] = column;
+  });
+  return configs;
+}
+
+/**
+ * Parse filter configs for FilterBox.
+ */
+function getFilterConfigsFromFilterBox(form_data = {}) {
   const {
     date_filter,
     filter_configs = [],
     show_druid_time_granularity,
     show_druid_time_origin,
     show_sqla_time_column,
     show_sqla_time_granularity,
+    table_filter,
   } = form_data;
-  let configs = filter_configs.reduce(
-    ({ columns, labels }, config) => {
-      let defaultValues = config.defaultValue;
-      // defaultValue could be ; separated values,
-      // could be null or ''
-      if (config.defaultValue) {
-        defaultValues = config.defaultValue.split(';');
-      }
-      const updatedColumns = {
-        ...columns,
-        [config.column]: config.vals || defaultValues,
-      };
-      const updatedLabels = {
-        ...labels,
-        [config.column]: config.label,
-      };
+  const configs = { columns: {}, labels: {} };
 
-      return {
-        columns: updatedColumns,
-        labels: updatedLabels,
-      };
-    },
-    { columns: {}, labels: {} },
-  );
+  filter_configs.forEach(({ column, label, defaultValue, vals }) => {
+    const defaultValues =
+      typeof defaultValue === 'string' ? defaultValue.split(';') : defaultValue;
+    configs.columns[column] = vals || defaultValues;
+    configs.labels[column] = label;
+  });
 
   if (date_filter) {
-    let updatedColumns = {
-      ...configs.columns,
-      [TIME_FILTER_MAP.time_range]: form_data.time_range,
-    };
-    const updatedLabels = {
-      ...configs.labels,
-      ...Object.entries(TIME_FILTER_MAP).reduce(
-        (map, [key, value]) => ({
-          ...map,
-          [value]: TIME_FILTER_LABELS[key],
-        }),
-        {},
-      ),
-    };
```

Comment:
`Reduce` is confusing. We should avoid `reduce`: https://twitter.com/jaffathecake/status/1213077702300852224

@graceguo-supercat

Reply:
I feel everyday  in the internet  there will someone claim one or another new theory...🤷‍♀️
I saw so many ppl object to his opinion. Did he offer any reason WHY he made claim?

Reply:
Reduce code is often very difficult to read and plain old for loops or forEach functions are almost always easier to understand.

Programs should favor simplicity and clarity, not succinctness. 

This guy works for Google and is in charge of promoting new browser features such as Service Workers.

Reply:
@ktmud Realize this is old, but wanted to add my 2cents. Performance considerations aside, I find reduce to frequently be faster to write and more readable than for loops and forEach. The key reasons being that I can cognitively follow the steps in sequence by knowing the goal is to turn A into Z, and knowing that somewhere along the chain, I must take that step to go from A to C. As soon as that becomes imperative, I lose the ability to follow it sequentially since variables, intermediary steps, etc. must be declared first, and instead have to make assumptions about what the purpose of the code is. In a similar vein, that's why I typically hate non-curried underscore / lodash, because it reverses the call order and makes you start at the innermost function call.

That being said, the real reason why we write code that way is to have the ability to pause / resume execution at any point in the chain (i.e. generators).

Reply:
Interestingly, I find reduce harder to read also because I was not able to follow it sequentially. It kind of places the input and output in a very twisted flow:

```js
input.reduce((intermediateOutput, input) => {
   return intermediateOutput
}, output);
```

Maybe I just didn't use it often enough, but every time I encounter a reduce function, I always have to think what the output look like and sometimes it could even be difficult to find where the output object was initialized.

For this specific case, the original code looks like this:

```js
    const updatedLabels  = {
      ...configs.labels,
      ...Object.entries(TIME_FILTER_MAP).reduce(
        (map, [key, value]) => ({
          ...map,
          [value]: TIME_FILTER_LABELS[key],
        }),
        {},
      ),
    };
    configs.labels = updatedLabels;
```

Without reduce, it's just:

```js
    Object.entries(TIME_FILTER_MAP).forEach(([key, column]) => {
      configs.labels[column] = TIME_FILTER_LABELS[key];
    });
```

Is reduce really faster to write and read?